components/scream/src/physics/shoc/atmosphere_macrophysics.hpp

```diff
@@ -102,9 +102,15 @@ class SHOCMacrophysics : public scream::AtmosphereProcess
       const auto sub_z_int = ekat::subview(z_int, i);
       const auto s_z_int = ekat::scalarize(sub_z_int);
       Kokkos::parallel_for(Kokkos::TeamThreadRange(team, nlev_packs), [&] (const Int& k) {
-        // Exner
+        const auto range = ekat::range<IntSmallPack>(k*Spack::n);
+        const Smask in_nlev_range = (range < nlev);
+
+        // Inverse of Exner. Assert that exner != 0 when in range before computing.
         const Spack p_mid_ik(p_mid(i,k));
-        exner(i,k)  = PF::exner_function(p_mid_ik);
+        const Spack exner_ik = PF::exner_function(p_mid_ik);
+        const Smask nonzero = (exner_ik != 0);
+        EKAT_KERNEL_ASSERT((nonzero || !in_nlev_range).any());
+        inv_exner(i,k).set(nonzero, 1.0/exner_ik);
```

Comment:
In the AD, we were passing `exner`, now we change the name to `inv_exner`, compute the inverse of `exner`, and pass this to `shoc_main`. Internally, `shoc_main` now matches in both cpp and f90.